### DISCUSSION TITLE
"exposed beyond app through Intent.getData()"

### DISCUSSION THREAD

**Discussion by pd4321** [beeware/toga#3038] [Q&A]
Hi,
I wanted to transform this part of the code for opening a PDF file in Windows to Android::

```
path_to_acrobat = os.path.abspath('c:\Program Files\Adobe\Acrobat DC\Acrobat\Acrobat.exe')
process = subprocess.Popen([path_to_acrobat, '/A',str(strana) , path_to_pdf], shell=False, `stdout=subprocess.PIPE)`
```

I compiled this

```
    async def openFileDialog(self, widget):
        from toga.platform import current_platform
        if current_platform == "android":
            from android.content import Intent
            from java.io import File

            try:
                file = File(str('/data/data/cz.pd.Level/files/chaquopy/AssetFinder/app/Level/resources/LEVEL 281.pdf'))
                intent = Intent(Intent.ACTION_VIEW)
                FileProvider.getUriForFile(MainActivity.this, BuildConfig.APPLICATION_ID + ".provider",file)
                intent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                result = await self.app._impl.intent_result(intent)

            except Exception as chyba:
                self.main_window.info_dialog(f"Chyba", f"{chyba}")
```

Instead of opening the file I get the error "exposed beyond app through Intent.getData()". Do you know what to do about it?

[They came up with something here. ](https://stackoverflow.com/questions/38200282/android-os-fileuriexposedexception-file-storage-emulated-0-test-txt-exposed?rq=1) But I don't understand how to use it in Toga/Python

In the future, I want to open it from internal memory/memory card. But now I don't want to deal with access rights issues, so I put the file in resouces.

Thanks
![Screenshot_20241203_003432_Level Index](https://github.com/user-attachments/assets/4105904c-c756-4e07-8122-eae0604a7a36)

**Comment by freakboy3742** [comment]
I'm not entirely clear on what it is you're trying to do. 

Your original code uses subprocess to start a completely separate process to open a file. That approach won't work on Android, because you can't use subprocesses.

However, File intents aren't really what you want either. A File intent is the analog of Python's `open()` method - that won't "open" a file in the sense of "loading it into Acrobat", it "opens" a file so it can be read on a byte-by-byte basis. 

It sounds like what you want is the [Action View Intent](https://developer.android.com/reference/android/content/Intent.html#ACTION_VIEW) - that's the mechanism you'd use to open a URL  in a web browser, AFAIK the same approach is used for *any* URL scheme or content type, and it's up to the device to determine how to satisfy that View request. 

However, that approach also *won't* work with a file in your app's resources, because that won't be a URI that is visible to another application (which is what you need to "open in acrobat" on Android). It also presumes that the user has an app installed that can handle PDF content (i.e., that they have acrobat, or an analog of it, installed).

**Comment by pd4321** [comment]
I need to display a PDF file in an application. The method doesn't matter. The first code is just an example of how I solve it in Windows. If it is possible without another application, then it is better. I tried to use Android resources, but without success.

```
from android.graphics.pdf import PdfRenderer

reader = PdfRenderer('resources/LEVEL 281.pdf')
pocet = reader.getPageCount()
self.main_window.info_dialog(f"PDF", f"Page count: {pocet}")
```

←[33mW/python.stderr: Error in async handler: No module named 'android'←[0m

And it might be necessary to insert ["parcelFileDescriptor"](https://developer.android.com/reference/android/graphics/pdf/PdfRenderer) somewhere.

**Reply by freakboy3742** [reply]
> ←[33mW/python.stderr: Error in async handler: No module named 'android'←[0m

That error is a little bit of a misdirection; unfortunately, when a Java import fails, it always reports as `android` being the missing module, even when the actual problem is deeper in the package structure.

In this case, it looks like `PdfRenderer` *should* be a legal import, so I'm not sure what is going on there. If you are able to provide a complete reproduction case and error log, we can take a closer look at the problem.

**Reply by pd4321** [reply]
Probably a syntax error. I think this will be the best solution.
![tgrd](https://github.com/user-attachments/assets/f5cb635c-32c1-48c9-9938-3b981c814d05)

**Reply by freakboy3742** [reply]
The page you've screenshotted is a road map. It's a collection of ideas we're thinking about. It doesn't mean we're currently implementing them, or that we have any plans to implement them. It just means it's an idea we're amenable to.

**Reply by pd4321** [reply]
I know. And this idea is the best.
